RickAndMorty-iOS/Views/Locations/LocationDetailsViewController.swift

```diff
@@ -0,0 +1,179 @@
+//
+//  Created by Calvin Pak on 2023-03-14.
+//
+
+import UIKit
+import Combine
+
+class LocationDetailsViewController: UIViewController {
+
+    enum LocationDetailsSection: Int, CaseIterable {
+        case info
+        case residents
+        case emptyInfo
+        case emptyResidents
+    }
+
+    weak var coordinator: MainCoordinator?
+    let locationDetailsView = LocationDetailsView()
+    let viewModel = LocationDetailsViewModel()
+    var locationId: String
+
+    typealias DataSource = UICollectionViewDiffableDataSource<LocationDetailsSection, AnyHashable>
+    typealias Snapshot = NSDiffableDataSourceSnapshot<LocationDetailsSection, AnyHashable>
+    private var dataSource: DataSource!
+    private var cancellables = Set<AnyCancellable>()
+    var snapshot = Snapshot()
+
+    init(locationId: String) {
+        self.locationId = locationId
+        super.init(nibName: nil, bundle: nil)
+    }
+
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    override func loadView() {
+        view = locationDetailsView
+        navigationController?.navigationBar.prefersLargeTitles = true
+        locationDetailsView.collectionView.delegate = self
+        locationDetailsView.collectionView.refreshControl = UIRefreshControl()
+        locationDetailsView.collectionView.refreshControl?.addTarget(self, action: #selector(onRefresh), for: .valueChanged)
+    }
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        configureDataSource()
+        showEmptyData()
+        subscribeToViewModel()
+        viewModel.locationId = locationId
+    }
+
+    func showEmptyData() {
+        snapshot.deleteAllItems()
+        snapshot.appendSections([.info, .residents, .emptyInfo, .emptyResidents])
+        snapshot.appendItems(Array(repeatingExpression: EmptyData(id: UUID()), count: 2), toSection: .emptyInfo)
+        snapshot.appendItems(Array(repeatingExpression: EmptyData(id: UUID()), count: 4), toSection: .emptyResidents)
+        self.dataSource.apply(snapshot, animatingDifferences: true)
+    }
+
+    func subscribeToViewModel() {
+        viewModel.location.sink(receiveValue: { [self] location in
+            self.title = location.name
+            if !location.residents.isEmpty {
+                snapshot.deleteAllItems()
+                snapshot.appendSections([.info, .residents])
+                snapshot.appendItems([LocationDetails(location), LocationDetails(location)], toSection: .info)
+                snapshot.appendItems(location.residents, toSection: .residents)
+                self.dataSource.apply(snapshot, animatingDifferences: true)
+            }
+            // Dismiss refresh control.
+            DispatchQueue.main.async {
+                self.locationDetailsView.collectionView.refreshControl?.endRefreshing()
+            }
+
+        }).store(in: &cancellables)
+    }
+
+    @objc func onRefresh() {
+        viewModel.locationId = self.locationId
```

Comment:
You should make function 'refresh' in ViewModel and then modify its values there.